### DISCUSSION TITLE
How do we train lora for Flux-Fill-Dev?

### DISCUSSION THREAD

**Discussion by matabear-wyx** [bghira/SimpleTuner#1180] [Q&A] [answered]
The newly released Flux-Fill-Dev (Inpainting/Outpainting) is incredible. However, the LoRA I previously trained on Flux-Dev doesn't work directly with Flux-Fill-Dev, so I'm planning to retrain a LoRA specifically for Flux-Fill-Dev.

From my quick analysis, Flux-Fill-Dev takes an input of a 16-channel image latent, a 16-channel masked image latent, and a 64-channel mask. My initial idea for input preparation is to use the 16-channel image latent repeated twice, along with a 64-channel mask filled entirely with zeros (black mask). 

Any thoughts or suggestions?

**Comment by chengyou-jia** [comment]
any new update for your training? i am also trying to train flux-fill-dev with lora.

**Reply by matabear-wyx** [reply]
Yes, you can try my method, it works properly

**Reply by chengyou-jia** [reply]
Thank you for your reply! How is the performance of the fill-model trained with LoRA? Would it be possible for you to share the code for the update? I would greatly appreciate it! 😊

**Reply by matabear-wyx** [reply, marked as answer]
What I did is really simple and raw, as I discussed:
`def pack_fill_latents(latents, batch_size, num_channels_latents, height, width):
    latents = latents.view(
        batch_size, num_channels_latents, height // 2, 2, width // 2, 2
    )
    latents = latents.permute(0, 2, 4, 1, 3, 5)
    latents = latents.reshape(
        batch_size, (height // 2) * (width // 2), num_channels_latents * 4
    )
    # Step 2: Repeat the packed latents
    repeated_latents = latents.repeat(1, 1, 2)  # Repeat along the channel dimension

    # Step 3: Create a black mask of the required shape
    mask = torch.zeros((batch_size, (height // 2) * (width // 2), 8 * 8 * 4), device=latents.device)

    # Step 4: Concatenate repeated latents with the black mask
    packed_fill_latents = torch.cat([repeated_latents, mask], dim=-1)
    
    return packed_fill_latents`

**Reply by chengyou-jia** [reply]
thanks, so did you prepare a dataset(using real masks)and test it's effectiveness?

**Reply by Sebastian-Zok** [reply]
Hey, this sounds promising. Would you mind sharing your whole code? I can't seem to get the changes running with validation

**Reply by matabear-wyx** [reply]
Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.

**Reply by Bilal143260** [reply]
> Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.

Which codebase are you exactly referring to? Is this the one for starting point ?
[https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)

**Reply by matabear-wyx** [reply]
> > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> 
> Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)

Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.

**Reply by Bilal143260** [reply]
Got it. Thanks for reply. I will share the results here. 

> > > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> > 
> > 
> > Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)
> 
> Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.

Got it. Thanks for the reply. I will share the results here.

**Reply by chengyou-jia** [reply]
Looking forward for your results !

**Reply by JesseBrown1980** [reply]
> Got it. Thanks for reply. I will share the results here.
> 
> > > > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> > > 
> > > 
> > > Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)
> > 
> > 
> > Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.
> 
> Got it. Thanks for the reply. I will share the results here.

how did you all load the data into the trainer?

**Reply by JesseBrown1980** [reply]
> > > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> > 
> > 
> > Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)
> 
> Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.

Yes!!!, how to load the data into the loader?

**Reply by Bilal143260** [reply]
> Got it. Thanks for reply. I will share the results here.
> 
> > > > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> > > 
> > > 
> > > Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)
> > 
> > 
> > Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.
> 
> Got it. Thanks for the reply. I will share the results here.

Here we go! basically I have trained a virtual try-off lora where you input an image and the model kinda extracts the garments from that and converts them into ghost images. Once again thanks for the suggestion @matabear-wyx 
<img width="715" alt="Screenshot 2024-12-23 at 1 24 38 PM" src="https://github.com/user-attachments/assets/10f07894-5413-4323-92f8-9588f528e75e" />

**Reply by CalvinHerbst** [reply]
I

> Yes, you can try my method, it works properly

Could you specify what you are using to train? This is great if you got it to work but it's not clear how to me. Thanks

**Reply by fenglincong** [reply]
I try to fine-tune black-forest-labs/FLUX.1-Fill-dev with the default configuration, but I encountered a bug(diffusers/examples/dreambooth/train_dreambooth_lora_flux.py)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (4096x64 and 384x3072)

**Reply by fenglincong** [reply]
> > Got it. Thanks for reply. I will share the results here.
> > > > > Yes, my co-worker used a real mask for a **repainting** task, and it worked properly. However, I suggest following the Diffusers codebase to minimize coding effort. As you can see, what I did here was simply use a dummy mask.
> > > > 
> > > > 
> > > > Which codebase are you exactly referring to? Is this the one for starting point ? [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_flux.py](url)
> > > 
> > > 
> > > Yes, and basically all you need is a custom DataLoader. Then just use methods from FluxFillPipeline.
> > 
> > 
> > Got it. Thanks for the reply. I will share the results here.
> 
> Here we go! basically I have trained a virtual try-off lora where you input an image and the model kinda extracts the garments from that and converts them into ghost images. Once again thanks for the suggestion @matabear-wyx <img alt="Screenshot 2024-12-23 at 1 24 38 PM" width="715" src="https://private-user-images.githubusercontent.com/118292770/399456064-10f07894-5413-4323-92f8-9588f528e75e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MTcyNDIsIm5iZiI6MTczODkxNjk0MiwicGF0aCI6Ii8xMTgyOTI3NzAvMzk5NDU2MDY0LTEwZjA3ODk0LTU0MTMtNDMyMy05MmY4LTk1ODhmNTI4ZTc1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwN1QwODI5MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YzRjNDc1NzZiOGU3OTY5OTBhMzEyMjNjZWE2NjM1ODA4MzUxZGNhNmNkYjY0OWFkZTIyODM4MzFmYjZhZjdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.K0ALZ8abTz7Tj1HrBDUjDw4kC-kdH_trbTSwuLcT-ho">

I try to fine-tune black-forest-labs/FLUX.1-Fill-dev with the default configuration, but I encountered a bug(diffusers/examples/dreambooth/train_dreambooth_lora_flux.py)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (4096x64 and 384x3072)

**Reply by fenglincong** [reply]
> # Step 3: Create a black mask of the required shape
> mask = torch.zeros((batch_size, (height // 2) * (width // 2), 8 * 8 * 4), device=latents.device)

Does 0(black) indicate the area to be repainted and 1(white) indicate the area to be retained?

**Reply by by1e11** [reply]
> > # Step 3: Create a black mask of the required shape
> > mask = torch.zeros((batch_size, (height // 2) * (width // 2), 8 * 8 * 4), device=latents.device)
> 
> Does 0(black) indicate the area to be repainted and 1(white) indicate the area to be retained?

Just the opposite.

**Comment by bc129697** [comment]
> The newly released Flux-Fill-Dev (Inpainting/Outpainting) is incredible. However, the LoRA I previously trained on Flux-Dev doesn't work directly with Flux-Fill-Dev, so I'm planning to retrain a LoRA specifically for Flux-Fill-Dev.
> 
> From my quick analysis, Flux-Fill-Dev takes an input of a 16-channel image latent, a 16-channel masked image latent, and a 64-channel mask. My initial idea for input preparation is to use the 16-channel image latent repeated twice, along with a 64-channel mask filled entirely with zeros (black mask).
> 
> Any thoughts or suggestions?

你好，我最近也在用flux-fill-dev作为基座模型来训练，数据输入的channel也是16+16+64，但是训练出来的结果很差，想问一下你训练得怎么样了

**Comment by bc129697** [comment]
> The newly released Flux-Fill-Dev (Inpainting/Outpainting) is incredible. However, the LoRA I previously trained on Flux-Dev doesn't work directly with Flux-Fill-Dev, so I'm planning to retrain a LoRA specifically for Flux-Fill-Dev.
> 
> From my quick analysis, Flux-Fill-Dev takes an input of a 16-channel image latent, a 16-channel masked image latent, and a 64-channel mask. My initial idea for input preparation is to use the 16-channel image latent repeated twice, along with a 64-channel mask filled entirely with zeros (black mask).
> 
> Any thoughts or suggestions?

Hello, I have recently been using flux-fill-dev as the base model for training. The data input channel is also 16+16+64, but the training results are very poor. I would like to ask about your training results.

**Reply by matabear-wyx** [reply]
Flux Fill is still a distilled model after all, which means it still can't learn new distributions. I'm not sure exactly what task you're working on, but generally speaking, I’d recommend training only specific layers of the LoRA instead of doing a full finetune.

**Reply by bghira** [reply]
distilled models can learn just fine.

**Comment by Sebastian-Zok** [comment]
Here's my [training script](https://github.com/Sebastian-Zok/FLUX-Fill-LoRa-Training).  I used just one instance prompt describing the inpaining task, instead of a caption for each sample. The LoRa adapted my concept with just a couple of hundred steps.  But there are some known issues: like the validation is broken and also the masking is hardcoded, but I hope it helps someone to adapt to their use cases :)

**Reply by juliuskuehn1** [reply]
I try to get your script running. On what hardware and settings have you trained the lora ?

**Reply by Sebastian-Zok** [reply]
I trained it on an A100 from Runpod.io using Prodigy

**Reply by juliuskuehn1** [reply]
A100 80 GB or 40 GB?

**Reply by Sebastian-Zok** [reply]
80GB, it goes up to around 65 as far as I remember 👌

**Reply by juliuskuehn1** [reply]
great thank you

**Reply by Painkillerzzz** [reply]
Hi Sebastian, I ran your training script and took the diffusers' dog examples as the instance data. I kept your instance prompt "A character turnaround 45-degreed to the left" and also used it while generating. However the result looks like this:
![flux-fill-dev](https://github.com/user-attachments/assets/27f44ac8-e4ee-4586-b44b-03ff66bd5d34)
I don't think it learned the feature of the dog. What do you think about it?

**Reply by Sebastian-Zok** [reply]
Hey :)
have you adapted the masking strategy?
I yet did not use the script to learn a character LoRa, just to have a these "pose transfer"  LoRas, where the AI learns to use the context of the unmasked part, so it won't have to learn the character features within its distribution. Here's a sample that was generated with a 300 step LoRa:
![Sample](https://github.com/user-attachments/assets/4d1d03e6-c027-485a-85a8-1fee40c73ed5)

I could imagine a character LoRa will need more steps, but I'm curious and will give it a try in the evening 👌

**Reply by Painkillerzzz** [reply]
Yes, I adapted your hardcoded masking strategy. Sorry I didn't realize that its only for pose transfer, but as my result posed above, the generated part (right) does not look like the reference part (left). I didn't modify much of your script, just added --gradient_checkpointing and removed the validate prompt during training. Do you know the possible reason of this?
training script:
```
export MODEL_NAME="black-forest-labs/FLUX.1-Fill-dev"
export INSTANCE_DIR="/workspace/diffusers/photos/dog"
export OUTPUT_DIR="/workspace/diffusers/trained-flux/inpaint-lora/dog"
export NCCL_P2P_DISABLE=1
export NCCL_IB_DISABLE=1

accelerate launch /workspace/diffusers/examples/dreambooth/train_dreambooth_inpaint_lora_flux.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="bf16" \
  --instance_prompt="A character turnaround 45-degreed to the left" \
  --resolution=1024 \
  --train_batch_size=1 \
  --guidance_scale=1 \
  --gradient_accumulation_steps=4 \
  --optimizer="prodigy" \
  --learning_rate=1. \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=500 \
  --seed="0" \
  --gradient_checkpointing
```
inference script:
```python
import torch
from diffusers import FluxFillPipeline
from diffusers.utils import load_image

image = load_image("https://huggingface.co/datasets/sebastianzok/validationImageAndMask/resolve/main/image.png")
mask = load_image("https://huggingface.co/datasets/sebastianzok/validationImageAndMask/resolve/main/mask.png")

pipe = FluxFillPipeline.from_pretrained("black-forest-labs/FLUX.1-Fill-dev", torch_dtype=torch.bfloat16).to("cuda")
pipe.load_lora_weights("/workspace/diffusers/trained-flux/inpaint-lora/dog", torch_dtype=torch.bfloat16)

image = pipe(
    prompt="A character turnaround 45-degreed to the left",
    image=image,
    mask_image=mask,
    height=1632,
    width=1232,
    guidance_scale=30,
    num_inference_steps=50,
    max_sequence_length=512,
    generator=torch.Generator("cpu").manual_seed(0)
).images[0]
image.save(f"flux-fill-dev.png")
```

**Reply by Sebastian-Zok** [reply]
The reason is that you did not use the same train dataset 😅 

Since the script got some attention, I made some improvements to it, replaced all the hard coded stuff and adapted the README. 
I also tested to train the dog LoRa and it come out to work okay (400 steps), as you can see:
![sample1](https://github.com/user-attachments/assets/d9098dd4-ac5d-4bf0-a83e-764e41b76ef8)

I guess the quality could be improved with a better dataset, since the dog example is very limited and has very big generation areas, that from my experience, I would say FLUX.1-Fill struggles with.  

I described everything in the README, so feel free to test it out for your use case. and if you have any questions, don't hesitate to ask :)

**Reply by Sebastian-Zok** [reply]
Maybe also check out this: https://huggingface.co/black-forest-labs/FLUX.1-Fill-dev/discussions/31

**Reply by Painkillerzzz** [reply]
Thanks for your reply! I realized that was a dumb question TAT. Yes I took the corgi dataset with the hard-coded half masking strategy and that makes no sense. I will try another dataset with suitable masks.

**Reply by Painkillerzzz** [reply]
Hi Sebastian, thank you again for your script and guidance. I still have two questions after trying to train flux dev inpainting.
1. I notice that guidance_scale=1 is set during training, however guidance_scale is usually set to 30~50 during generating. I'm a little confused about this difference of setting during the two processes.
2. The learning_rate is set to 1 in training, I think it is still a high lr even for lora training, but the result is good actually. (The default parameter for [SD Dreambooth lora inpainting](https://github.com/huggingface/diffusers/blob/main/examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint_lora.py) is 5e-6.) I wonder if you might know some empirical conclusions about this setup.

**Reply by Sebastian-Zok** [reply]
Hey,
1. I don't know the exact details, but most of the training scripts "disable" guidance through setting it to 1 during training (see kohya or the other diffuser scripts). After that you can increase it for interference, while here a low guidance scale should do the work, since you have the LoRa that should help to do the job beside the prompt. 
2. The learning rate is set to 1 due to the Prodigy Optimizer, which dynamically adjusts the learning rate based on the previous gradient  [(See more here)](https://huggingface.co/blog/sdxl_lora_advanced_script#adaptive-optimizers`), while the script you linked is using Adam.

**Reply by Painkillerzzz** [reply]
This is very helpful and I understand these settings better now. Thank you so much!

**Reply by Sheldongg** [reply]
Why can the flux dev model be trained on consumer-grade GPUs, while flux fill requires high-end GPUs like the A100? Is this due to the model architecture itself or are there some optimization techniques involved?

**Reply by Sebastian-Zok** [reply]
Optimization, so it should be possible to adapt SimpleTuner or similar for this, but you need to have data loading for the masks, which would change some code in the project. Also I once stumbled on someone who claimed to train on consumer hardware, but didn't wanted to share the code.
In the end the diffuser script was much easier for me to adapt the changes to, I also tried it on SimpleTuner but failed 😅

Edit: Actually the someone has now shared [their training code](https://github.com/lrzjason/T2ITrainer/blob/main/doc%2Fflux_fill.md) and it seems to run on 24GB VRAM. 
I will give it a try in the afternoon (edit: tried it but the script thrown an error after the first epoch, do not have time to investigate)
